### PR TITLE
Improved SSL Configurations

### PR DESCRIPTION
**Pull request by cycomachead**
This (re-) cleans up the SSL configurations and enables directing all non-https traffic (except for Snap!) to the HTTPS domain. 

It also sets the Secure parameter on the cooking specifying it should only be used with the https domain. 

There are other mostly minor (in comparison) security improvements, like setting a HSTS header, and enabling "perfect forward secrecy" which reduces the risk of man in the middle attacks.

### CODE REVIEW

File: app.lua [hunk 1]
```diff
@@ -43,51 +43,18 @@ package.loaded.config = require("lapis.config").get()
 
 local app = package.loaded.app
 
+-- Store whitelisted domains
```
Comment (cycomachead) [inline app.lua]: This file belongs in a separate PR

File: config.lua [hunk 2]
```diff
@@ -45,7 +45,8 @@ config('development', {
 
     -- development needs no special SSL or cert config.
     primary_nginx_config = 'locations.conf',
-    secondary_nginx_config = 'locations.conf'
+    -- A comment disables an include.
```
Comment (cycomachead) [inline config.lua]: fix comment

File: config.lua [hunk 3]
```diff
@@ -65,14 +66,14 @@ config({'production', 'staging'}, {
 config('production', {
     site_name = 'Snap Cloud',
     num_workers = 8,
-    primary_nginx_config = 'berkeley-production.conf',
-    secondary_nginx_config = 'cs10.org-production.conf'
+    primary_nginx_config = 'http-only.conf',
+    secondary_nginx_config = 'include nginx.conf.d/ssl-production.conf;'
```
Comment (cycomachead) [inline config.lua]: add note about needing a `;`

File: cors.lua [hunk 4]
```diff
@@ -0,0 +1,63 @@
+-- CORS module
+-- ==========
```
Comment (cycomachead) [inline cors.lua]: fix

File: nginx.conf [hunk 5]
```diff
@@ -40,18 +38,14 @@ http {
 
     # We use a separate server block for each host to serve specific certs.
     server {
-        server_name ${{HOSTNAME}};
+        server_name ${{HOSTNAME}} ${{SECONDARY_HOSTNAME}};
```
Comment (cycomachead) [inline nginx.conf]: Update secondary hostname in dev. (Cosmetic.)

File: nginx.conf.d/berkeley-production.conf [hunk 6]
```diff
@@ -1,8 +0,0 @@
-# Configuration for Berkeley hosts
```
Comment (cycomachead) [inline nginx.conf.d/berkeley-production.conf]: These files were condensed from 4 files to 2.

One file for production and one for staging. It seems easier to maintain and understand. They still share relevant configuration via includes.

File: nginx.conf.d/http-only.conf [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+# Snap! is exempted from SSL so people can use Robots.
+location /snap/ {
+    sub_filter <head> '<head>\n\t<meta name="snap-cloud-domain" location="https://$host">';
+    alias snap/;
+}
+
+# Everything else should have SSL.
+location / {
+ return 301 https://$host$request_uri;
+}
```
Comment (cycomachead) [inline nginx.conf.d/http-only.conf]: fix all line endings.

File: nginx.conf.d/http-only.conf [hunk 8]
```diff
@@ -0,0 +1,18 @@
+# nginx amplify server monitoring agent
+# a local deamon connets to this endpoint so it does not need SSL
+location /nginx_status {
+    stub_status on;
+    allow 127.0.0.1;
+    deny all;
+}
+
+# Snap! is exempted from SSL so people can use Robots.
+location /snap/ {
```
Comment (cycomachead) [inline nginx.conf.d/http-only.conf]: Add comment http-only snap should still use https api

File: nginx.conf.d/ssl-staging.conf [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+# Configuration blocks for cloud.snap.berkeley.edu and snap-cloud.cs10.org
+server {
+    # Staging should not be avaible to everyone, because the data aren't safe.
+    auth_basic           'Snap! Cloud Staging';
+    auth_basic_user_file .htpasswd;
+
+    listen 443 ssl http2;
+    server_name staging.snap.berkeley.edu;
+
+    ssl_certificate     certs/cloud_snap_berkeley_edu.cer;
```
Comment (cycomachead) [inline nginx.conf.d/ssl-staging.conf]: fixme. We have staging certs now.

File: nginx.conf.d/ssl-staging.conf [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+    ssl_certificate_key certs/cloud_snap_berkeley_edu.key;
+
+    include nginx.conf.d/locations.conf;
+}
+
+server {
+    # Staging should not be avaible to everyone, because the data aren't safe.
+    # auth_basic           'Snap! Cloud Staging';
```
Comment (cycomachead) [inline nginx.conf.d/ssl-staging.conf]: make sure to re-enable when done testing.

File: nginx.conf.d/ssl.conf [hunk 11]
```diff
@@ -1,5 +0,0 @@
-# Shared SSL settings for all hosts.
```
Comment (cycomachead) [inline nginx.conf.d/ssl.conf]: moved to ssl-shared.

File: nginx.conf [hunk 12]
```diff
@@ -29,8 +29,6 @@ http {
     lua_shared_dict auto_ssl 1m;
     lua_shared_dict auto_ssl_settings 64k;
     lua_shared_dict page_cache 15m;
-    ssl_session_cache   shared:SSL:10m;
```
Comment (cycomachead) [inline nginx.conf]: moved to SSL shared config file -- there's no functional difference, just logically grouping settings.

File: nginx.conf [hunk 13]
```diff
@@ -27,11 +27,7 @@ events {
 
 http {
     include nginx.conf.d/mime.types;
-    lua_shared_dict auto_ssl 1m;
```
Comment (cycomachead) [inline nginx.conf]: These are not needed since we aren't using auto-ssl anymore. 

The two below were moved to ssl-shared.conf so they're with all the other ssl settings.

**Comment by bromagosa** [comment]
This is great. Should we pull all these to production this week? I've time on Wednesday, and I'd wait for you to wake up (or begin when you're still awake, what do you prefer?)

**Comment by cycomachead** [comment]
That sounds good! 

I’ll do a review of the production server setup before weds, but it _should_ be as simple as pulling the repo and restarting. 

I generated the dhparam cert already, so it’s already in the right spot. (It takes about 5-10 minutes to generate for some reason, but it should only need to be done once.)

-- 
Michael Ball
From my iPhone
michaelball.co
